pkg/reconciler/channel/channel.go

```diff
@@ -0,0 +1,412 @@
+	}
+	// If there's nothing to update, just return.
+	if reflect.DeepEqual(channel.Status, desired.Status) {
+		return channel, nil
+	}
+	becomesReady := desired.Status.IsReady() && !channel.Status.IsReady()
+	// Don't modify the informers copy.
+	existing := channel.DeepCopy()
+	existing.Status = desired.Status
+
+	ch, err := c.RunClientSet.EventsV1alpha1().Channels(desired.Namespace).UpdateStatus(existing)
+	if err == nil && becomesReady {
+		duration := time.Since(ch.ObjectMeta.CreationTimestamp.Time)
+		c.Logger.Infof("Channel %q became ready after %v", channel.Name, duration)
+
+		if err := c.StatsReporter.ReportReady("Channel", channel.Namespace, channel.Name, duration); err != nil {
+			logging.FromContext(ctx).Infof("failed to record ready for Channel, %v", err)
+		}
+	}
+
+	return ch, err
+}
+
+func (c *Reconciler) syncSubscribers(ctx context.Context, channel *v1alpha1.Channel) error {
+	if channel.Status.Subscribers == nil {
+		channel.Status.Subscribers = []eventingduck.SubscriberStatus(nil)
+	}
+
+	subCreates := []eventingduck.SubscriberSpec(nil)
+	subUpdates := []eventingduck.SubscriberSpec(nil)
+	subDeletes := []eventingduck.SubscriberStatus(nil)
+
+	// Make a map of name to PullSubscription for lookup.
+	pullsubs := make(map[string]pubsubv1alpha1.PullSubscription)
+	if subs, err := c.getPullSubscriptions(ctx, channel); err != nil {
+		c.Logger.Infof("Failed to list PullSubscriptions, %s", err)
+	} else {
+		for _, s := range subs {
+			pullsubs[s.Name] = s
+		}
+	}
+
+	exists := make(map[types.UID]eventingduck.SubscriberStatus)
+	for _, s := range channel.Status.Subscribers {
+		exists[s.UID] = s
+	}
+
+	if channel.Spec.Subscribable != nil {
+		for _, want := range channel.Spec.Subscribable.Subscribers {
+			if got, ok := exists[want.UID]; !ok {
+				// If it does not exist, then update it.
+				subCreates = append(subCreates, want)
+			} else {
+				_, found := pullsubs[resources.GenerateSubscriptionName(want.UID)]
+				// If did not find or the PS has updated generation, update it.
+				if !found || got.ObservedGeneration != want.Generation {
+					subUpdates = append(subUpdates, want)
+				}
+			}
+			// Remove want from exists.
+			delete(exists, want.UID)
+		}
+	}
+
+	// Remaining exists will be deleted.
+	for _, e := range exists {
+		subDeletes = append(subDeletes, e)
+	}
+
+	for _, s := range subCreates {
+		genName := resources.GenerateSubscriptionName(s.UID)
+		c.Logger.Infof("Channel %q will create subscription %s", channel.Name, genName)
+
+		ps := resources.MakePullSubscription(&resources.PullSubscriptionArgs{
+			Owner:      channel,
+			Name:       genName,
+			Project:    channel.Spec.Project,
+			Topic:      channel.Status.TopicID,
+			Secret:     channel.Spec.Secret,
+			Labels:     resources.GetPullSubscriptionLabels(controllerAgentName, channel.Name, genName),
+			Subscriber: s,
+		})
+		ps, err := c.RunClientSet.PubsubV1alpha1().PullSubscriptions(channel.Namespace).Create(ps)
+		if err != nil {
```

Comment:
Might want to check the error case of "already exists" if something created it already (name collision) and check the ownerref and if it's not us, propagate.

Reply:
I will try this out in the e2e tests and see what it does. It will error at the moment and that error will be logged but not as nice as you want.